File: src/main/java/com/smnas/backend/entity/Subject.java
package com.smnas.backend.entity;


import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import javax.persistence.*;
import java.time.LocalDateTime;
import java.util.List;

@Entity
@Table(name = "subject")
@NoArgsConstructor
@AllArgsConstructor
@Data
public class Subject {
    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    @Column(name = "id")
    private Long id;

    @Column(name = "subject_name")
    private String subjectName;

    @ManyToMany
    @JoinTable(
            name = "subject_teacher",
            joinColumns = @JoinColumn(name = "subject_id"),
            inverseJoinColumns = @JoinColumn(name = "teacher_id"))
    private List<User> teachers;

    @Column(name = "created_at")
    @CreationTimestamp
    private LocalDateTime createdAt;

    @Column(name = "updated_at")
    @UpdateTimestamp
    private LocalDateTime updatedAt;
}

File: client/src/types/subject.ts
import {UserResponse} from "./user";
import {TeacherResponse} from "./teacher";

export interface SubjectRequest {
    subjectName: string,
    teacherIds: number[],
}

export interface SubjectResponse {
    id: number,
    subjectName: string,
    teachers: UserResponse[],
    createdAt: Date,
    updatedAt: Date
}

export interface TeacherSubjectRequest {
    subjectId: number,
    teacherId: number
}

export interface TeacherSubjectResponse {
    id: number,
    teacher: TeacherResponse,
    subject: SubjectResponse
}